### ISSUE TITLE
How to add `orml_tokens` benchmarks to the runtime?

### ISSUE THREAD

**Issue by Chralt98** [open-web3-stack/open-runtime-module-library#871]
Hey, I face the following error message, when trying to add the `orml_tokens` pallet to the benchmarks:

> no function or associated item named `run_benchmark` found for struct `orml_tokens::Pallet` in the current scope

I activated the benchmark feature in the runtime Cargo.toml 

```
runtime-benchmarks = [
...
  "orml-tokens/runtime-benchmarks",
...
]
```

I added `orml_tokens` here:

```rust
#[cfg(feature = "runtime-benchmarks")]
mod benches {
	define_benchmarks!(
                ...
		[orml_tokens, Tokens]
	);
}
```

What am I missing? In addition to that I think `orml_nft` and `orml_currencies` have no benchmark feature activated, so using the `type WeightInfo = ();` for `orml_currencies` is accurate in the runtime?

**Comment by Chralt98** [comment]
I am using `polkadot-v0.9.29` currently.

**Comment by Chralt98** [comment]
Oh I see, that you use default weights, nice! https://github.com/open-web3-stack/open-runtime-module-library/blob/80ed32dad524955b63e2c3fea7c14defcd24d79c/tokens/src/weights.rs#L40

**Comment by Chralt98** [comment]
Do those default weights have any flaws, which should be taken into consideration to benchmark them manually?

**Comment by Chralt98** [comment]
Could you provide the benchmarks for each pallet, so that the users can just benchmark it quicker without writing the benchmarks manually but with weights based on their own reference machine? 

Because currently the default weights force the users to use the recommended reference machine, right?

**Comment by shaunxw** [comment]
You would want to run your own benchmark for each orml pallets used, as different runtime configs could have very different results. For instance, the hooks functions impl in `Config` has impact on some specific dispatch-able calls. Refer to config in Acala runtime as an example of how to config benchmarks for orml pallets https://github.com/AcalaNetwork/Acala/blob/master/runtime/acala/src/lib.rs

**Comment by xlc** [comment]
It is hard (if not impossible) to write generic benchmarking code. The frame-benchmarking framework is not generic to handle that. We use our fork orml-benchmkaring to be able to write runtime specific benchmarks and therefore able to actually correctly do it without tons of overhead on trying to make it generic.

So yeah it is not ideal but you can just use Acala benchmarking code as reference and make it your own one https://github.com/AcalaNetwork/Acala/blob/master/runtime/mandala/src/benchmarking/tokens.rs

Feel free to make suggestion if you can think of a better way.

**Comment by Chralt98** [comment]
Thanks for your new input! Loved it.

**Comment by Chralt98** [comment]
I get the following error when adding the `orml_tokens` benchmarks:

```
  error[E0573]: expected type, found module `crate::benchmarking::orml_tokens::benches`
     --> /Users/userproject/container-chains/templates/simple/runtime/src/lib.rs:654:23
      |
  654 |         [orml_tokens, crate::benchmarking::orml_tokens::benches]
      |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not a type
  ...
  762 |             list_benchmarks!(list, extra);
      |             ----------------------------- in this macro invocation
      |
      = note: this error originates in the macro `list_benchmarks` (in Nightly builds, run with -Z macro-backtrace for more info)
      
      ....
        error[E0573]: expected type, found module `crate::benchmarking::orml_tokens::benches`
     --> /Users/userproject/container-chains/templates/simple/runtime/src/lib.rs:654:23
      |
  654 |         [orml_tokens, crate::benchmarking::orml_tokens::benches]
      |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not a type
  ...
  816 |             add_benchmarks!(params, batches);
      |             -------------------------------- in this macro invocation
      |
      = note: this error originates in the macro `add_benchmarks` (in Nightly builds, run with -Z macro-backtrace for more info)
```

My folder structure is `/benchmarking/orml_tokens/benches` as outlined. It looks like `frame_benchmarking::define_benchmarks` requires a type rather than a module like in my case. But this is basically the same approach as on the Acala main branch [here](https://github.com/AcalaNetwork/Acala/blob/3b389e5df3f00c6c8a257742cd5089e0464c47b2/runtime/mandala/src/benchmarking/tokens.rs).

How can you in Acala use `benchmarking::tokens` inside `define_benchmarks` [here](https://github.com/AcalaNetwork/Acala/blob/3b389e5df3f00c6c8a257742cd5089e0464c47b2/runtime/acala/src/lib.rs#L1932), which is a module defined [here](https://github.com/AcalaNetwork/Acala/blob/3b389e5df3f00c6c8a257742cd5089e0464c47b2/runtime/acala/src/benchmarking/mod.rs#L107) (not a type).

**Comment by xlc** [comment]
are you using the `define_benchmarks` from orml? can you share the code?

**Comment by Chralt98** [comment]
Oh, I used the `define_benchmarks` from `frame-benchmarking`. You got me, thanks for helping out!

Another option would be to use the `define_benchmarks!` macro for all pallets other than the orml pallets and then use `orml_benchmarking::{list_benchmark as orml_list_benchmark}` and `orml_benchmarking::{add_benchmark as orml_add_benchmark}` for each orml pallet separately. I did this in combination of using `add_benchmarks!` and `list_benchmarks!` for all other pallets.
